pkg/proxy/proxy.go

```diff
@@ -391,18 +392,52 @@ func (p *Proxy) addUserContext() echo.MiddlewareFunc {
 				return next(ctx)
 			}
 
-			userID, username, err := p.extractUserID(ctx.Request())
+			token, err := p.extractUserToken(ctx.Request())
 			if err != nil {
 				return crterrors.NewUnauthorizedError("invalid bearer token", err.Error())
 			}
-			ctx.Set(context.SubKey, userID)
-			ctx.Set(context.UsernameKey, username)
+			ctx.Set(context.SubKey, token.Subject)
+			ctx.Set(context.UsernameKey, token.PreferredUsername)
+			ctx.Set(context.EmailKey, token.Email)
 
 			return next(ctx)
 		}
 	}
 }
 
+// ensureUserIsNotBanned rejects the request if the user is banned
+func (p *Proxy) ensureUserIsNotBanned() echo.MiddlewareFunc {
+	return func(next echo.HandlerFunc) echo.HandlerFunc {
+		return func(ctx echo.Context) error {
+			errorResponse := func(err *crterrors.Error) error {
+				ctx.Logger().Error(errs.Wrap(err, "workspace list error"))
+				ctx.Response().Writer.Header().Set("Content-Type", "application/json")
+				ctx.Response().Writer.WriteHeader(int(err.Code))
+				return json.NewEncoder(ctx.Response().Writer).Encode(err.Status)
+			}
+
+			email := ctx.Get(context.EmailKey).(string)
+			if email == "" {
+				return errorResponse(crterrors.NewUnauthorizedError("unauthenticated request", "anonymous access is not allowed"))
```

Comment:
I wonder why you actually sets the response here (headers, response code, etc) instead of returning `crterrors.NewUnauthorizedError("unauthenticated request", "anonymous access is not allowed")` here as is as we do in other middlewares like `addUserContext`? I didn't check the code but apparently it works for `addUserContext`?